File: backend/src/BuildERP.Infrastructure/Services/SettingsService.cs
using BuildERP.Application.Common.Interfaces;
using BuildERP.Application.Features.AuditLogs;
using BuildERP.Application.Features.Settings;
using BuildERP.Domain.Entities;
using BuildERP.Infrastructure.Persistence;
using Microsoft.EntityFrameworkCore;
using System.Text.Json;

namespace BuildERP.Infrastructure.Services;

public class SettingsService : ISettingsService
{
    private readonly ApplicationDbContext _db;
    private readonly IAuditLogService _auditLog;
    private readonly ICurrentUserService _currentUser;

    public SettingsService(ApplicationDbContext db, IAuditLogService auditLog, ICurrentUserService currentUser)
    {
        _db = db;
        _auditLog = auditLog;
        _currentUser = currentUser;
    }

    public async Task<CompanySettingsDto> GetAsync(CancellationToken ct = default)
    {
        var settings = await _db.CompanySettings.FirstOrDefaultAsync(ct);
        if (settings == null)
        {
            settings = new CompanySettings { CompanyName = "ConstructPro", Currency = "PKR" };
            _db.CompanySettings.Add(settings);
            await _db.SaveChangesAsync(ct);
        }
        return MapToDto(settings);
    }

    public async Task<CompanySettingsDto> UpdateAsync(UpdateCompanySettingsRequest request, CancellationToken ct = default)
    {
        var settings = await _db.CompanySettings.FirstOrDefaultAsync(ct);
        if (settings == null)
        {
            settings = new CompanySettings();
            _db.CompanySettings.Add(settings);
        }

        var oldValues = JsonSerializer.Serialize(new { settings.CompanyName, settings.Currency, settings.NTN });

        settings.CompanyName = request.CompanyName; settings.Address = request.Address;
        settings.Phone = request.Phone; settings.Email = request.Email; settings.Website = request.Website;
        settings.NTN = request.NTN; settings.STRN = request.STRN; settings.Currency = request.Currency;
        settings.FinancialYearStart = request.FinancialYearStart; settings.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync(ct);

        await _auditLog.LogAsync(new AuditLogEntry
        {
            UserId = _currentUser.UserId,
            UserEmail = _currentUser.Email ?? "",
            Action = "Update",
            EntityType = "CompanySettings",
            OldValues = oldValues,
            NewValues = JsonSerializer.Serialize(new { settings.CompanyName, settings.Currency, settings.NTN }),
        }, ct);

        return MapToDto(settings);
    }

    private static CompanySettingsDto MapToDto(CompanySettings s) => new()
    {
        Id = s.Id, CompanyName = s.CompanyName, Address = s.Address,
        Phone = s.Phone, Email = s.Email, Website = s.Website,
        NTN = s.NTN, STRN = s.STRN, LogoPath = s.LogoPath,
        Currency = s.Currency, FinancialYearStart = s.FinancialYearStart, UpdatedAt = s.UpdatedAt
    };
}

File: backend/src/settings/settings.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { UpdateSettingsDto } from './dto/settings.dto';

@Injectable()
export class SettingsService {
  constructor(private readonly prisma: PrismaService) {}

  async getSettings() {
    let settings = await this.prisma.companySettings.findFirst();

    if (!settings) {
      settings = await this.prisma.companySettings.create({
        data: {
          companyName: 'ConstructPro',
          currency: 'PKR',
        },
      });
    }

    return settings;
  }

  async updateSettings(dto: UpdateSettingsDto) {
    const existing = await this.prisma.companySettings.findFirst();

    const data: any = {};

    if (dto.companyName !== undefined) data.companyName = dto.companyName;
    if (dto.address !== undefined) data.address = dto.address;
    if (dto.phone !== undefined) data.phone = dto.phone;
    if (dto.email !== undefined) data.email = dto.email;
    if (dto.website !== undefined) data.website = dto.website;
    if (dto.ntn !== undefined) data.ntn = dto.ntn;
    if (dto.strn !== undefined) data.strn = dto.strn;
    if (dto.currency !== undefined) data.currency = dto.currency;
    if (dto.financialYearStart !== undefined) {
      data.financialYearStart = dto.financialYearStart ? new Date(dto.financialYearStart) : null;
    }

    if (existing) {
      return this.prisma.companySettings.update({
        where: { id: existing.id },
        data,
      });
    } else {
      return this.prisma.companySettings.create({
        data: {
          companyName: dto.companyName ?? 'ConstructPro',
          currency: dto.currency ?? 'PKR',
          ...data,
        },
      });
    }
  }
}
